File: src/evidence.ts
import type { PaperRecord } from "./literature.js";
import { verifiedRecords } from "./literature.js";

export type EvidenceGate = {
  related_work_verified: boolean;
  strong_baseline_defined: boolean;
  dataset_defined: boolean;
  metric_defined: boolean;
  claim_evidence_mapped: boolean;
};

export function submissionReady(gate: EvidenceGate): boolean {
  return Object.values(gate).every(Boolean);
}

export function blockingReasons(gate: EvidenceGate): string[] {
  const reasons: string[] = [];
  if (!gate.related_work_verified) reasons.push("verified_related_work_missing");
  if (!gate.strong_baseline_defined) reasons.push("strong_baseline_missing");
  if (!gate.dataset_defined) reasons.push("dataset_missing");
  if (!gate.metric_defined) reasons.push("metric_missing");
  if (!gate.claim_evidence_mapped) reasons.push("claim_evidence_missing");
  return reasons;
}

export function evaluateEvidenceGate(
  papers: PaperRecord[] = [],
  options: {
    baselines?: string[];
    datasets?: string[];
    metrics?: string[];
    claimEvidenceRows?: Array<Record<string, unknown>>;
  } = {}
): EvidenceGate {
  const rows = options.claimEvidenceRows ?? [];
  const mapped = rows.some(structuredEvidenceRow);
  return {
    related_work_verified: verifiedRecords(papers).length > 0,
    strong_baseline_defined: (options.baselines ?? []).some(verifiedText),
    dataset_defined: (options.datasets ?? []).some(verifiedText),
    metric_defined: (options.metrics ?? []).some(verifiedText),
    claim_evidence_mapped: mapped
  };
}

export function evidenceGateMarkdown(gate: EvidenceGate): string {
  return `# Evidence Gate

- Submission readiness: ${submissionReady(gate) ? "ready" : "blocked"}
- Blocking reasons: ${blockingReasons(gate).join(", ") || "none"}

| Gate | Passed |
| --- | --- |
| Verified related work | ${yesNo(gate.related_work_verified)} |
| Strong baseline defined | ${yesNo(gate.strong_baseline_defined)} |
| Dataset defined | ${yesNo(gate.dataset_defined)} |
| Metric defined | ${yesNo(gate.metric_defined)} |
| Claim evidence mapped | ${yesNo(gate.claim_evidence_mapped)} |

This gate is evidence-based. Generated revised-plan prose does not make a project
submission-ready until the artifacts above contain verified evidence.
`;
}

function yesNo(value: boolean): string {
  return value ? "yes" : "no";
}

function verifiedText(value: unknown): boolean {
  if (typeof value !== "string") return false;
  const normalized = value.trim().toLowerCase();
  if (!normalized) return false;
  return !["todo", "tbd", "placeholder", "planned", "unknown", "unspecified"].some((marker) => normalized.includes(marker));
}

function stringValue(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function structuredEvidenceRow(row: Record<string, unknown>): boolean {
  const confidence = typeof row.confidence === "number" ? row.confidence : Number.NaN;
  return Boolean(
    verifiedText(row.paper_id) &&
      verifiedText(row.claim) &&
      isEvidenceClaimType(row.claim_type) &&
      Number.isFinite(confidence) &&
      confidence > 0 &&
      confidence <= 1 &&
      Number.isFinite(Number(row.page)) &&
      Number(row.page) >= 1 &&
      verifiedText(row.quote) &&
      verifiedText(row.chunk_id) &&
      ["verified", "measured"].includes(stringValue(row.status)) &&
      hasPdfChunkProvenance(row.provenance)
  );
}

function hasPdfChunkProvenance(value: unknown): boolean {
  if (!value || typeof value !== "object") return false;
  const provenance = value as Record<string, unknown>;
  return provenance.source === "pdf_chunk" && verifiedText(provenance.artifact) && verifiedText(provenance.extracted_at);
}

function isEvidenceClaimType(value: unknown): boolean {
  return typeof value === "string" && ["method", "dataset", "metric", "baseline", "limitation", "result", "threat", "future_work"].includes(value);
}

File: idea2repo/evidence.py
"""Evidence gates for CCF-A readiness."""

from __future__ import annotations

from dataclasses import dataclass

from .literature import PaperRecord, verified_records


@dataclass(frozen=True)
class EvidenceGate:
    """Evidence-backed readiness state separate from idea potential scoring."""

    related_work_verified: bool
    strong_baseline_defined: bool
    dataset_defined: bool
    metric_defined: bool
    claim_evidence_mapped: bool

    @property
    def submission_ready(self) -> bool:
        return all(
            (
                self.related_work_verified,
                self.strong_baseline_defined,
                self.dataset_defined,
                self.metric_defined,
                self.claim_evidence_mapped,
            )
        )

    @property
    def blocking_reasons(self) -> tuple[str, ...]:
        reasons: list[str] = []
        if not self.related_work_verified:
            reasons.append("verified_related_work_missing")
        if not self.strong_baseline_defined:
            reasons.append("strong_baseline_missing")
        if not self.dataset_defined:
            reasons.append("dataset_missing")
        if not self.metric_defined:
            reasons.append("metric_missing")
        if not self.claim_evidence_mapped:
            reasons.append("claim_evidence_missing")
        return tuple(reasons)


def evaluate_evidence_gate(
    papers: list[PaperRecord] | None = None,
    *,
    baselines: list[str] | None = None,
    datasets: list[str] | None = None,
    metrics: list[str] | None = None,
    claim_evidence_rows: list[dict[str, str]] | None = None,
) -> EvidenceGate:
    verified = verified_records(papers or [])
    rows = claim_evidence_rows or []
    mapped_rows = [
        row for row in rows
        if _verified_text(row.get("claim", ""))
        and _verified_text(row.get("required_evidence", ""))
        and _verified_text(row.get("planned_artifact", ""))
        and row.get("status") in {"verified", "measured"}
    ]
    return EvidenceGate(
        related_work_verified=len(verified) > 0,
        strong_baseline_defined=bool([item for item in baselines or [] if _verified_text(item)]),
        dataset_defined=bool([item for item in datasets or [] if _verified_text(item)]),
        metric_defined=bool([item for item in metrics or [] if _verified_text(item)]),
        claim_evidence_mapped=bool(mapped_rows),
    )


def evidence_gate_markdown(gate: EvidenceGate) -> str:
    status = "ready" if gate.submission_ready else "blocked"
    reasons = ", ".join(gate.blocking_reasons) or "none"
    return f"""# Evidence Gate

- Submission readiness: {status}
- Blocking reasons: {reasons}

| Gate | Passed |
| --- | --- |
| Verified related work | {_yes_no(gate.related_work_verified)} |
| Strong baseline defined | {_yes_no(gate.strong_baseline_defined)} |
| Dataset defined | {_yes_no(gate.dataset_defined)} |
| Metric defined | {_yes_no(gate.metric_defined)} |
| Claim evidence mapped | {_yes_no(gate.claim_evidence_mapped)} |

This gate is evidence-based. Generated revised-plan prose does not make a project
submission-ready until the artifacts above contain verified evidence.
"""


def _yes_no(value: bool) -> str:
    return "yes" if value else "no"


def _verified_text(value: str) -> bool:
    normalized = " ".join(value.split()).casefold()
    if not normalized:
        return False
    placeholders = ("todo", "tbd", "placeholder", "planned", "unknown", "unspecified")
    return not any(marker in normalized for marker in placeholders)
